validation/validate.ts

```diff
@@ -0,0 +1,265 @@
+ * @param page - Page containing the nodes in nodeRefs
+ */
+async function removeRedundantNodes(
+  nodeRefs: NodeRef[],
+  page: Page
+): Promise<NodeRef[]> {
+  const redundantNodes: NodeRef[] = [];
+  for (const nodeRefA of nodeRefs) {
+    for (const nodeRefB of nodeRefs) {
+      // Any node that has an ancestor in nodeRefs is considered redundant because
+      // outerHTML includes all descendants.
+      const isRedundant = await page.evaluate(
+        (nodeA, nodeB) => nodeA.contains(nodeB),
+        nodeRefA.handle,
+        nodeRefB.handle
+      );
+      // nodeA contains nodeA, so we must ensure that nodeRefA !== nodeRefB or all
+      // nodes will be considered redundant.
+      if (isRedundant && nodeRefA !== nodeRefB) {
+        redundantNodes.push(nodeRefB);
+      }
+    }
+  }
+  return nodeRefs.filter(nodeRef => !redundantNodes.includes(nodeRef));
+}
+
+/**
+ * Gets all nodes used by Chrome to compute the accessible name for nodeRef.
+ * @param nodeRef - Node whose accessible name is being computed.
+ * @param client - CDPSession for page.
+ * @param page - Page containing nodeRef.
+ */
+async function getNodesUsedByChrome(
+  nodeRef: NodeRef,
+  client: CDPSession,
+  page: Page
+): Promise<NodeRef[]> {
+  const stack: NodeRef[] = [];
+  const nodesUsed: NodeRef[] = [];
+  // Track backendIds of visited nodes to avoid infinite cycle.
+  const visitedNodes: Protocol.DOM.BackendNodeId[] = [];
+  stack.push(nodeRef);
+  // Iterative DFS traverses nodes connected by label references
+  while (stack.length > 0) {
+    const currentNodeRef = stack.pop();
+    nodesUsed.push(currentNodeRef!);
+
+    const axTree = (await client.send('Accessibility.getPartialAXTree', {
+      backendNodeId: currentNodeRef!.backendId,
+    })) as Protocol.Accessibility.GetPartialAXTreeResponse;
+
+    // Check if any AXNodes descandant of currentNodeRef are labelled
+    for (const axNode of axTree.nodes) {
+      let labelNodes: Protocol.Accessibility.AXRelatedNode[] = [];
+      const sources: Protocol.Accessibility.AXValueSource[] =
+        axNode.name?.sources ?? [];
+
+      for (const source of sources) {
+        if (source.type === 'relatedElement') {
+          // Handles nodes connected by attribute value (aria-labelleby)
+          if (source.attributeValue?.relatedNodes) {
+            labelNodes = source.attributeValue.relatedNodes;
+            // Handles nodes connected natively (<label>)
+          } else if (source.nativeSourceValue?.relatedNodes) {
+            labelNodes = source.nativeSourceValue.relatedNodes;
+          }
+        }
+      }
+
+      // Repeat the process for all unvisited label nodes.
+      for (const labelNode of labelNodes) {
+        const labelNodeRef = await getNodeRefFromBackendId(
+          labelNode.backendDOMNodeId,
+          client,
+          page
+        );
+        if (labelNodeRef && !visitedNodes.includes(labelNodeRef.backendId)) {
```

Comment:
It'd be more efficient to do the `visitedNodes` check before getting the `NodeRef`